### ISSUE TITLE
Exceptiona are thrown after tests execution, if  the database test container is closed prematurely.

### ISSUE THREAD

**Issue by stoyants** [jdbi/jdbi#2629]
Hi guys,
I found out that after execution of TestScript in the "jdbi3-oracle12" some Exceptions pop-at the end.
Tests are fine, they are all successful, but these Exceptions are confusing and annoying.
An example can be found in the attached file.
[TestScript.log](https://github.com/jdbi/jdbi/files/14233909/TestScript.log)

I have dig a little bit and found out, that the problem happen as follows:
In the TestcontainersDatabaseInformationSupplier class the main thread "run()" method has an "infinite" loop, which is creating a new TestcontainersDatabaseInformation (a new schema and/or user, depending on the database) as soon as one is consumed via there "get()" method. This includes also execution of the list of statements for schema/user initialization in the private executeStatements(....) method.
So, the problem happens if the test is so short and fast, that it finishes, before the next schema / user is initialised completely.
The main testing thread invokes the close() method, it does not check are any statements are currently executed and then container shutdown is initiated. Then statements can't be executed anymore in the "[database-schema-creator" and the java.sql.SQLRecoverableException is thrown.

I also have a suggestion how this can be avoided.
1. as first and simple step, the loop executing initializing statement can check is the TestcontainersDatabaseInformationSupplier is already close. like this:

```java
    private void executeStatements(final List<String> statements) throws SQLException {
        try (Connection c = dataSource.getConnection()) {
            for (String statement : statements) {
                if (closed.get()) {
                    break;
                }
                try (Statement stmt = c.createStatement()) {
                    stmt.executeUpdate(statement);
                }
            }
        }
    }
```

2. But the above is still not sufficient, we have to protect the invocation of the "executeStatements" using ReentrantLock mutex.

```java
    private final ReentrantLock mutex = new ReentrantLock();

    @Override
    public void close() {
        LOG.info("Shutdown initiated...");
        if (!this.closed.getAndSet(true)) {
            mutex.lock();
            executor.shutdownNow();
            ...
        }
    }

    @Override
    public void run() {
        while (!closed.get()) {
            try {
                ......
                try {
                    mutex.lock();
                    executeStatements(databaseInformation.getCreationScript());
                } finally {
                    mutex.unlock();
                }
                ......
            }
        }
        stopped.countDown();
    }

```

Of course, will be glad to see any any other suggestions how to avoid this Exception.

If you agree to review my suggestion, I can open a MR.

Best Regards, Stoyan

**Comment by hgschmie** [comment]
Hi @stoyants,

Thank you for looking into this! Your analysis is correct. What happens is that the test framework has independent references to the TestContainer extension and the Jdbi extension. A typical stanza in a testcontainers test looks like this:

```java
final class PostgisPluginTest {

    @Container
    public static JdbcDatabaseContainer<?> pgContainer = new PostgreSQLContainer<>(DockerImageName.parse("postgis/postgis:13-3.3-alpine")
        .asCompatibleSubstituteFor("postgres"));

    @RegisterExtension
    public JdbiExtension pgExtension = JdbiTestcontainersExtension.instance(pgContainer)
...
}
```

When Junit shuts down the extensions, it does not know that those two are related and that the JdbiExtension may do an operation on the JdbcDatabaseContainer. It shuts down the extensions unconditionally and in undefined order, so the database container may suddenly "disappear" from the JdbiExtension. This causes those exceptions.

As you rightfully pointed out, these exceptions are harmless in this context (they are caused by the drivers when the connection goes away while there is still an operation pending). Coordination would require collaboration of the test framework (startup and shutdown are controlled by JUnit). Adding `close()` checks or a mutex does not help, as the shutdown happens outside the scope of the Jdbi extension.

The very same exception would not be harmless if they happen e.g. in the middle of a regular test if the database crashes. From the PoV of the JdbiExtension, crashing the database and having it shut down by the test framework are indistinguishable. 

It should be possible to avoid this by using the `@Order` annotation on the extensions, e.g. 

```java
final class MyTest {

    @Container
    @Order(1)
    public static JdbcDatabaseContainer<?> dbContainer = ...

    @RegisterExtension
    @Order(2)
    public JdbiExtension jdbiExtension = JdbiTestcontainersExtension.instance(dbContainer)
...
}
```

This should ensure that the database container is started before the jdbi extension and stopped after the it. I have not tried that (see https://junit.org/junit5/docs/current/user-guide/#extensions-registration-programmatic for more details).

**Comment by stoyants** [comment]
Hi @hgschmie,

Thanks for the long and comprehensive reply.

I carefully read your explanations and agree with most of your statements.

But I would like to mention the following:

@Order annotation alone is not solving the problem. I tested it, added also @TestMethodOrder(MethodOrderer.OrderAnnotation.class) to the class, but problem persists.
Honestly, I can't imagine how JUnit alone can use any other order, as first JdbcDatabaseContainer is static and is passed as a parameter to the second JdbiExtension extension.

What solves the problem is to make the JdbiExtension also statuc. In fact, why not, the test class will be anyway instantiated only once. Then somehow (magic?) problem dissapears. I don't have a any explanation of this.

But I agree with @stevenschlansker that developers are much happier with a framework which doesn't impose requirements on them and simply allows them to get the job done as quickly as possible.

I'm not sure I understand your statement that "mutex doesn't help". 
The "database-schema-creator" thread protects execution of statements in "executeStatements" by locking the mutex.
The "close()" method is called synchronously by JUnit from the "main" thread, in JdbiTestcontainersExtension.stopExtension(...), so that "mutex.lock()" waits there until executeStatements() is finished and mutex is released.
I have implemented, tested it and it works as expected. In fact I tested even by puting intentinal delays between statements in "executeStatements" -  still works. 
The "break" in "executeStatements()" is just to be nice to the JUnit and to respect its request to close as soon as we can.
Please have a look at the following PR: https://github.com/jdbi/jdbi/pull/2632

Best Regards, Stoyan

**Comment by hgschmie** [comment]
> It's a little awkward to require test authors to manage two different extension declarations and order them. In practice, this isn't likely to happen correctly all the time.
> 
> Since the usual use case is that a test author is trying to accomplish one "thing" (get a Jdbi with a database backing it), can we create an extension that combines these two into one? Then you can opt in to advanced behavior (configure db and jdbi separately) while having the 99% use case ("I just want it to go") not require advanced configuration.

I hear you but that is the nature of testcontainers. 

In fact, we already do the same thing with the embedded postgres:

```java
    @RegisterExtension
    public static EmbeddedPgExtension pg = SingleDatabaseBuilder.instanceWithDefaults().build();

    @RegisterExtension
    public JdbiExtension postgres = JdbiExtension.postgres(pg);
```

We already use two separate extensions here. The reason for that is lifecycle:

- The database extension (EmbeddedPgExtension or the test container) are a class member (static field). Which means, there is one instance of that extension brought up when the test is set up and then torn down when all the unit tests in the class have been executed. E.g. for Oracle with a test class that contains 20 tests, that is the difference between 30 seconds and 10 minutes of test execution time.
- The Jdbi extension however is an instance field, so it is created and destroyed for each test invocation. Because there needs to be a new instance with a new Connection object and assorted friends (Handle, Jdbi) for each test. 

There is no way around this (and I tried when I wrote the testcontainer code). And it is especially tricky as testcontainers does not use the `@RegisterExtension` annotation but a meta-annotation (`@Container`) on the field *and* another annotation (`@Testcontainers`) on the class.

Is that ideal? Probably not. But if someone has *decided* to use testcontainers in their unit test, they have all of that already set up. For Jdbi, this only adds a single, instance scoped extension that leverages the testcontainer infrastructure.

Personally, I very much prefer the "jdbi native" solutions for postgres, h2 and sqlite. Supporting testcontainers is a service to the users that are using it (and spotify gave money to us and testcontainers at the same time, so I would not be surprised to learn that they use these two things together...).

**Comment by hgschmie** [comment]
> Hi @hgschmie,

Hi,

Thanks for taking time to look at my reply. The reason why the JdbiExtension is an instance field and the testcontainer is a class field is described here: https://junit.org/junit5/docs/current/user-guide/#extensions-registration-programmatic. The test container (which holds a reference to the database) should exist during the full lifecycle of the test while the test class itself (!) and the jdbi extension are instantiated for every test. Add e.g.

```java
    @BeforeEach
    public void debugSetup() {
        System.out.printf("container is %x, extension is %x, this is %x", dbContainer.hashCode(), extension.hashCode(), this.hashCode());
    }
```

to any of the tests in org.jdbi.v3.testing.junit5.tc. It will report e.g.

container is 345d053b, extension is 3d0cac1f, this is d5ce97f
container is 345d053b, extension is 1b9d9a2b, this is 659f226a
container is 345d053b, extension is 37348491, this is 1727e03a

So for all three unit tests in the same test class, the container stays constant, the extension is instantiated every time and the test class itself (!) is also instantiated every time. 




> 
> Thanks for the long and comprehensive reply.
> 
> I carefully read your explanations and agree with most of your statements.
> 
> But I would like to mention the following:
> 
> @order annotation alone is not solving the problem. I tested it, added also @TestMethodOrder(MethodOrderer.OrderAnnotation.class) to the class, but problem persists. Honestly, I can't imagine how JUnit alone can use any other order, as first JdbcDatabaseContainer is static and is passed as a parameter to the second JdbiExtension extension.

There is a difference between instantiation (which is done in the code) and the lifecycle (which is described here: https://junit.org/junit5/docs/current/user-guide/#extensions-lifecycle-callbacks and executed in the order as defined there). 


> What solves the problem is to make the JdbiExtension also static. In fact, why not, the test class will be anyway instantiated only once. 

That is not true. See above. The test class is instantiated separately for each test. 

> Then somehow (magic?) problem dissappears. I don't have a any explanation of this.

You can do that (it is explicitly supported in the extension code). Now you share the instance between all your tests (see above). You can not (!) run tests in parallel, at least not if you use the `JdbiExtension#getSharedHandle()` method, which is, as the name implies, shared. :-) 

You will also share database state (there is no new database for each of your test methods). 

There are actually good reasons to do so. If you have a test suite where one test sets up the database state and you lock down the execution order of the tests with `@Order` and they rely on each other, you can do that with the `JdbiExtension`. 

> 
> But I agree with @stevenschlansker that developers are much happier with a framework which doesn't impose requirements on them and simply allows them to get the job done as quickly as possible.

I sympathize with you but you need to take that up with the JUnit team. JUnit 5 is a great testing framework and its extensibility is unparalleled (try JUnit 4 or TestNG and you see what I mean). The current code is the best that I can write to integrate with JUnit 5, Jdbi and get the performance and the convenience of running tests fast and in isolation. Patches welcome.


> I'm not sure I understand your statement that "mutex doesn't help". The "database-schema-creator" thread protects execution of statements in "executeStatements" by locking the mutex. 

What the mutex does is straightforward: It blocks the thread that calls "close()" from making progress while the creator thread is busy executing statements against the database. But all the close code does is this:

```java
public void close() {
    LOG.info("Shutdown initiated...");
    if (!this.closed.getAndSet(true)) {
        executor.shutdownNow();
        try {
            if (!stopped.await(10, TimeUnit.SECONDS)) {
                LOG.warn("Could not shut down database creation thread within 10 seconds");
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        LOG.info("Shutdown completed.");
    }
}
```

So it sets the `closed` atomic to true and shuts the executor pool down. The former signals to the loop that it should not do another round but exit. This will terminate the creator thread in case it was e.g executing `nextSchema.put(databaseInformation);` and then jumping to the top of the while loop. The latter sends an Interrupt to all threads in the threadpool. If the creator thread does any I/O or is parked, it will receive the interrupt as an "InterruptedException", end up in the `catch(InterruptedException e)` block, reset the Interrupt status and exit the while block.

The exception trace that you posted looks like this:

```
[database-schema-creator] ERROR org.jdbi.v3.testing.junit5.tc.TestcontainersDatabaseInformationSupplier - SQL Exception caught:
java.sql.SQLRecoverableException: ORA-18730: Interrupted IO error.: Socket read interrupted
https://docs.oracle.com/error-help/db/ora-18730/
        at oracle.jdbc.driver.T4CStatement.executeForRows(T4CStatement.java:1126)
        at oracle.jdbc.driver.OracleStatement.executeSQLStatement(OracleStatement.java:1652)
        at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1417)
        at oracle.jdbc.driver.OracleStatement.executeUpdateInternal(OracleStatement.java:2194)
        at oracle.jdbc.driver.OracleStatement.executeLargeUpdate(OracleStatement.java:2153)
        at oracle.jdbc.driver.OracleStatement.executeUpdate(OracleStatement.java:2137)
        at oracle.jdbc.driver.OracleStatementWrapper.executeUpdate(OracleStatementWrapper.java:318)
        at com.zaxxer.hikari.pool.ProxyStatement.executeUpdate(ProxyStatement.java:119)
        at com.zaxxer.hikari.pool.HikariProxyStatement.executeUpdate(HikariProxyStatement.java)
        at org.jdbi.v3.testing.junit5.tc.TestcontainersDatabaseInformationSupplier.executeStatements(TestcontainersDatabaseInformationSupplier.java:128)
        at org.jdbi.v3.testing.junit5.tc.TestcontainersDatabaseInformationSupplier.run(TestcontainersDatabaseInformationSupplier.java:88)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        Suppressed: java.sql.SQLRecoverableException: ORA-17008: Closed connection
https://docs.oracle.com/error-help/db/ora-17008/
                at oracle.jdbc.driver.PhysicalConnection.requireOpenConnection(PhysicalConnection.java:12015)
                at oracle.jdbc.driver.PhysicalConnection.needLine(PhysicalConnection.java:4225)
                at oracle.jdbc.driver.OracleStatement.closeOrCache(OracleStatement.java:1930)
                at oracle.jdbc.driver.OracleStatement.close(OracleStatement.java:1913)
                at oracle.jdbc.driver.OracleStatementWrapper.close(OracleStatementWrapper.java:155)
                at com.zaxxer.hikari.pool.ProxyStatement.close(ProxyStatement.java:75)
                at org.jdbi.v3.testing.junit5.tc.TestcontainersDatabaseInformationSupplier.executeStatements(TestcontainersDatabaseInformationSupplier.java:127)
                ... 6 more
Caused by: java.io.InterruptedIOException: Socket read interrupted
        at oracle.net.nt.TimeoutSocketChannel.doSelect(TimeoutSocketChannel.java:706)
        at oracle.net.nt.TimeoutSocketChannel.doSelectForRead(TimeoutSocketChannel.java:599)
        at oracle.net.nt.TimeoutSocketChannel.read(TimeoutSocketChannel.java:501)
        at oracle.net.ns.NSProtocolNIO.doSocketRead(NSProtocolNIO.java:1228)
        at oracle.net.ns.NIOPacket.readHeader(NIOPacket.java:271)
        at oracle.net.ns.NIOPacket.readPacketFromSocketChannel(NIOPacket.java:204)
        at oracle.net.ns.NIOPacket.readFromSocketChannel(NIOPacket.java:147)
        at oracle.net.ns.NIOPacket.readFromSocketChannel(NIOPacket.java:120)
        at oracle.net.ns.NIONSDataChannel.readDataFromSocketChannel(NIONSDataChannel.java:105)
        at oracle.jdbc.driver.T4CMAREngineNIO.prepareForUnmarshall(T4CMAREngineNIO.java:893)
        at oracle.jdbc.driver.T4CMAREngineNIO.unmarshalUB1(T4CMAREngineNIO.java:464)
        at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:949)
        at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:456)
        at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:482)
        at oracle.jdbc.driver.T4CStatement.doOall8(T4CStatement.java:170)
        at oracle.jdbc.driver.T4CStatement.executeForRows(T4CStatement.java:1117)
        ... 15 more
```

The interesting piece is here: `Suppressed: java.sql.SQLRecoverableException: ORA-17008: Closed connection https://docs.oracle.com/error-help/db/ora-17008/ at oracle.jdbc.driver.PhysicalConnection.requireOpenConnection(PhysicalConnection.java:12015)`

So this is *not* an interrupted exception. This comes from the driver which says "oops, database connection is closed". And my interpretation is that this happens because the *server* was shut down while the thread was executing statements (there is a `org.jdbi.v3.testing.junit5.tc.TestcontainersDatabaseInformationSupplier.executeStatements(TestcontainersDatabaseInformationSupplier.java:128` in there). And that would be consistent with the testcontainer being shut down before the jdbi extension (which is supported by https://junit.org/junit5/docs/current/user-guide/#extensions-registration-programmatic-order that claims that "By default, extensions registered programmatically via @RegisterExtension or declaratively via @ExtendWith on fields will be ordered using an algorithm that is deterministic but intentionally nonobvious.").


> The "close()" method is called synchronously by JUnit from the "main" thread, in JdbiTestcontainersExtension.stopExtension(...), so that "mutex.lock()" waits there until executeStatements() is finished and mutex is released. 

Correct. But that may be after the fact (after the main thread has shut down the test container extension). If you look at the log file, there is a test that was executed successfully:

```
[main] INFO com.zaxxer.hikari.HikariDataSource - jdbi-template-pool (*.*) - Starting...
[main] WARN com.zaxxer.hikari.util.DriverDataSource - Registered driver with driverClassName=oracle.jdbc.driver.OracleDriver was not found, trying direct instantiation.
[main] INFO com.zaxxer.hikari.pool.HikariPool - jdbi-template-pool (*.*) - Added connection oracle.jdbc.driver.T4CConnection@6b6b4177
[main] INFO com.zaxxer.hikari.HikariDataSource - jdbi-template-pool (*.*) - Start completed.
[main] INFO com.zaxxer.hikari.HikariDataSource - jdbi-test-pool (jihchyujsyrb.izpdobdcolom) - Starting...
[main] WARN com.zaxxer.hikari.util.DriverDataSource - Registered driver with driverClassName=oracle.jdbc.driver.OracleDriver was not found, trying direct instantiation.
[main] INFO com.zaxxer.hikari.pool.HikariPool - jdbi-test-pool (jihchyujsyrb.izpdobdcolom) - Added connection oracle.jdbc.driver.T4CConnection@20bd3082
[main] INFO com.zaxxer.hikari.HikariDataSource - jdbi-test-pool (jihchyujsyrb.izpdobdcolom) - Start completed.
[main] INFO org.jdbi.v3.testing.junit5.tc.TestcontainersDatabaseInformationSupplier - Shutdown initiated...
[main] INFO org.jdbi.v3.testing.junit5.tc.TestcontainersDatabaseInformationSupplier - Shutdown completed.
[main] INFO com.zaxxer.hikari.HikariDataSource - jdbi-test-pool (jihchyujsyrb.izpdobdcolom) - Shutdown initiated...
[main] INFO com.zaxxer.hikari.HikariDataSource - jdbi-test-pool (jihchyujsyrb.izpdobdcolom) - Shutdown completed.
[main] INFO com.zaxxer.hikari.HikariDataSource - jdbi-template-pool (*.*) - Shutdown initiated...
[main] INFO com.zaxxer.hikari.HikariDataSource - jdbi-template-pool (*.*) - Shutdown completed.
```

This is the log from a successful test (start template-pool, start test pool, shut down test pool, shut down template pool). The last test in this test suite throws the exception. I would be curious to see if you can get this exception from another (not the last test in the test class) test. 

> I have implemented, tested it and it works as expected. In fact I tested even by putting intentional delays between statements in "executeStatements" - still works. The "break" in "executeStatements()" is just to be nice to the JUnit and to respect its request to close as soon as we can. Please have a look at the following PR: #2632

I will. I am not opposed to the change but I want to see evidence that it actually fixes the problem, not just change timing.

IMHO, the fix is much simpler and that is to change the catch statements to look like this:

```java
...
} catch (SQLException e) {
    if (!closed.get()) {
        LOG.error("SQL Exception caught:", e);
    }
    break; // while
}
```

which makes the errors go away in case they are thrown when the extension is shutting down.

**Comment by stoyants** [comment]
@hgschmie 
Thanks again for your time explaining all in such details. You are right about everything, including my wrong assumption about single instance of the test class - yes, it is instantiated for each test method, my bad, sorry.   I realized this when I was already in the bed and intended to correct myself, but you was faster :)

But please, look into the PR, it is really small. If you see any flaw in my logic, please let me know - I will be more than happy to improve it. And yes, you last suggestion - catching and processing an SQLException is always a good idea. If you want, I can add this to the PR as well. But it will only register the problem on a better way, will not avoid it.

Best regards, Stoyan

**Comment by hgschmie** [comment]
(to wrap this up: the messages that you see are actually not exceptions. They are logged by the connection pool, which is why they start with

```
[database-schema-creator] WARN com.zaxxer.hikari.pool.ProxyConnection - jdbi-template-pool (*.*) - Connection oracle.jdbc.driver.T4CConnection@4886b6bd marked as broken because of SQLSTATE(08006), ErrorCode(18730)
```

They come from here: https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java#L178

and then get rethrown which then creates the second, actual exception:

```
[database-schema-creator] ERROR org.jdbi.v3.testing.junit5.tc.TestcontainersDatabaseInformationSupplier - SQL Exception caught:
java.sql.SQLRecoverableException: ORA-18730: Interrupted IO error.: Socket read interrupted
```

suppressing logging will fix the second but not the first. Those can not be suppressed easily.

**Comment by stoyants** [comment]
Hi @hgschmie 
> Is that log from a "vanilla" run with jdbi3-oracle12 or did you change the code that produced that log?
Yes, "vanilla" :
![image](https://github.com/jdbi/jdbi/assets/17534482/819ecaaa-b69c-43e4-8d6d-c156388a4bb3)
and the log is attached.
[TestScript.log](https://github.com/jdbi/jdbi/files/14270966/TestScript.log)

Might be related to my configuration (Windows 10 + WSL)

You know, for each test which is expected to use N schemas the extension creates N+1. Yes, most of the cases N is 1, so for each test there are 2 schemas created. The last created schema is never used.
In this particular test (TestScript) JUnit 5 decide to execute the "long" test method "testIssue2021" first
The quick test method "testStatementParsing" is second, it starts, schema is created and consumed by the test, a second schema creations is initiated in the "database-schema-creator" thread, But test is so fast (no any SQL in fact, it only performs parsing) that it is over, complete test class is over and JUnit starts closing the containers. The Oracle gets shut-down, while creation of second schema is ongoing in the "database-schema-creator" thread and this generates the exceptions.

Best Regards, Stoyan

**Comment by hgschmie** [comment]
Yes, that is intentional. The "next schema" is prepared by the database preparer thread while the test is running. 

Instead of 

```
database schema prep -> | -> test -> database schema prep -> | -> test`
```

it is supposed to do

```
database schema prep -> | -> database schema prep -> | -> database schema prep
                        | -> test                    | -> test
```

If the schema prep is quick and the test runs long (which is a common case), the total test time is the upper bound. If the schema prep is slow and the tests are quick, then the upper bound is just schema creation time. 

Creating the "n+1th" schema is either fast in which case it does not matter or slow, in which case shutting down the creator thread should interrupt it (and the main thread waits at most 10 seconds for the shutdown).

**Comment by hgschmie** [comment]
> Hi @hgschmie
> 
> > Is that log from a "vanilla" run with jdbi3-oracle12 or did you change the code that produced that log?
[...]
> Might be related to my configuration (Windows 10 + WSL)

Thanks for redoing this. Weird. There is definitely logging in there that is not from our code. Do you happen to have any environment variables set. Especially this

```
2024-02-13 19:53:10 INFO I:thread-1 main entering args (oracle.jdbc.internal.AbstractConnectionBuilder$1@5ea419b1)
2024-02-13 19:53:10 INFO U:thread-1 main traceId=2A28BF25.
2024-02-13 19:53:10 INFO U:thread-1 main Session Attributes:
sdu=8192, tdu=2097152
nt: host=localhost, port=64211, socketOptions={0=YES, 1=NO, 17=0, 18=false, 2=30000, 20=true, 38=TLS, 23=40, 24=50, 40=false, 25=0}
    socket=null
client profile={oracle.net.encryption_types_client=(), oracle.net.crypto_seed=, oracle.net.authentication_services=(), oracle.net.setFIPSMode=false, oracle.net.kerberos5_mutual_authentication=false, oracle.net.encryption_client=ACCEPTED, oracle.net.crypto_checksum_client=ACCEPTED, oracle.net.crypto_checksum_types_cli
ent=()}
connection options=[host=localhost port=64211 protocol=TCP service_name=xepdb1 addr=(ADDRESS=(PROTOCOL=TCP)(HOST=localhost)(PORT=64211)) conn_data=(DESCRIPTION=(ADDRESS=(PROTOCOL=TCP)(PORT=64211)(HOST=127.0.0.1)(HOSTNAME=localhost))(CONNECT_DATA=(CID=(PROGRAM=ForkedBooter)(HOST=Lenovo-P51s)(USER=stoyants))(SERVICE_NA
ME=xepdb1))) done=true]
onBreakReset=false, dataEOF=false, negotiatedOptions=0x841, connected=false
2024-02-13 19:53:10 INFO U:thread-1 main traceId=2A28BF25, anoEnabled=true.
2024-02-13 19:53:10 INFO U:thread-1 main Got Resend, SessionTraceId = 2A28BF25
2024-02-13 19:53:10 INFO U:thread-1 main Connection established. Cleared conn strategy stack
2024-02-13 19:53:10 INFO I:thread-1 main CONNECTION_ID=z7OgxqMvSwmuXM/Z7lKXRg==,TENANT=XEPDB1,returning void
2024-02-13 19:53:10 INFO U:thread-48 database-schema-creator CONNECTION_ID=z7OgxqMvSwmuXM/Z7lKXRg==,TENANT=XEPDB1,null
```

seems to be internal tracing (all our test code uses slf4j and slf4j-simple as logging. This is clearly something else (java util logging most likely).

The exceptions you see may also be from the reader thread receiving an Interrupt (it is a `java.sql.SQLRecoverableException: ORA-18730: Interrupted IO error.: Socket read interrupted` after all). But those are normal conditions when shutting down a thread. Logging them is interesting but it does not constitute an error. The only really annoying one is the 

```
[database-schema-creator] WARN com.zaxxer.hikari.pool.ProxyConnection - jdbi-template-pool (*.*) - Connection oracle.jdbc.driver.T4CConnection@2a28bf25 marked as broken because of SQLSTATE(08006), ErrorCode(18730)
``` 

line, which is the database pool trying to be "helpful". See https://github.com/brettwooldridge/HikariCP/issues/2172 :-)

**Comment by stoyants** [comment]
Hi again, @hgschmie 

> 2024-02-13 19:53:10 INFO I:thread-1 main entering args (oracle.jdbc.internal.AbstractConnectionBuilder$1@5ea419b1)
2024-02-13 19:53:10 INFO U:thread-1 main traceId=2A28BF25.
2024-02-13 19:53:10 INFO U:thread-1 main Session Attributes:
sdu=8192, tdu=2097152
nt: host=localhost, port=64211, socketOptions={0=YES, 1=NO, 17=0, 18=false, 2=30000, 20=true, 38=TLS, 23=40, 24=50, 40=false, 25=0}
    socket=null

This above is "oracle stuff". Their driver is not open source, but [SDU/PDU](https://www.dba-oracle.com/phys_sdu_tdu.htm) are Oracle abbreviations.

Also,. Oracle Jdbc driver is using plain "java.util.logging", so they can rely only on the JRE and avoid any dependencies.

More important info is that I went to a different computer, plain Linux, even weaker CPU. If I would say that problem never appear, it will not be true. But it appeared like 3 times out of 50 or more test runs. Seems to me that in the setup that I "standardly" use (Win + WSL) the request to "executor.shutdown()" is not performed "instantly". I will investigate this now.

Best Regards, Stoyan

**Comment by hgschmie** [comment]
I think your problems *should* be gone with 3.45.0 and if they are not, you can now tweak with the shutdown time setting on the testcontainer extension.

Thank you for the discussion; user feedback is what makes Jdbi better and I found this very helpful and illuminating. As a postgres user myself, I rarely encounter problems like this.

**Comment by stoyants** [comment]
Hallo @hgschmie 

Yes, I can confirm, that the problem with unexpected error messages and stack-traces disappeared in 3.45.0
Thanks for this.

BTW, I have posted a question https://github.com/jdbi/jdbi/discussions/2647 regarding this logic for schema pre-creation.
And attached 2 tests. I hope it will be interesting to have a look there.

Best regards, Stoyan
